be/src/exec/spill/mem_table.cpp

```diff
@@ -98,34 +189,56 @@ Status OrderedMemTable::append_selective(const Chunk& src, const uint32_t* index
     Chunk* current = _chunk.get();
     size_t mem_usage = current->memory_usage();
     _chunk->append_selective(src, indexes, from, size);
+    _num_rows += size;
     mem_usage = current->memory_usage() - mem_usage;
 
     _tracker->consume(mem_usage);
     COUNTER_ADD(_spiller->metrics().mem_table_peak_memory_usage, mem_usage);
     return Status::OK();
 }
 
-Status OrderedMemTable::flush(FlushCallBack callback) {
+Status OrderedMemTable::finalize(workgroup::YieldContext& yield_ctx) {
+    SCOPED_TIMER(_spiller->metrics().mem_table_finalize_timer);
+    if (!_sort_done) {
+        SCOPED_RAW_TIMER(&yield_ctx.time_spent_ns);
+        // do sort
+        ASSIGN_OR_RETURN(_chunk, _do_sort(_chunk));
+        _chunk_slice.reset(_chunk);
+    }
+    RETURN_OK_IF_NEED_YIELD(yield_ctx.wg, &yield_ctx.need_yield, yield_ctx.time_spent_ns);
+    // seriealize data, store result into _block
+    auto& serde = _spiller->serde();
+    if (_block == nullptr) {
+        _block = std::make_shared<MemoryBlock>();
+    }
+    SerdeContext serde_ctx;
     while (!_chunk_slice.empty()) {
+        SCOPED_RAW_TIMER(&yield_ctx.time_spent_ns);
         auto chunk = _chunk_slice.cutoff(_runtime_state->chunk_size());
-        RETURN_IF_ERROR(callback(chunk));
+        RETURN_IF_ERROR(serde->serialize_to_block(serde_ctx, chunk, _block));
+        RETURN_OK_IF_NEED_YIELD(yield_ctx.wg, &yield_ctx.need_yield, yield_ctx.time_spent_ns);
     }
+    TRACE_SPILL_LOG << fmt::format("finalize spillable ordered memtable done, rows[{}], size[{}]", num_rows(),
+                                   _block->size());
+    // clear all data
     _chunk_slice.reset(nullptr);
-    int64_t consumption = _tracker->consumption();
-    _tracker->release(consumption);
-    COUNTER_ADD(_spiller->metrics().mem_table_peak_memory_usage, -consumption);
+    int64_t old_consumption = _tracker->consumption();
+    int64_t new_consumption = _block->size();
+    _tracker->set(new_consumption);
+    COUNTER_ADD(_spiller->metrics().mem_table_peak_memory_usage, new_consumption - old_consumption);
     _chunk.reset();
     return Status::OK();
 }
 
-Status OrderedMemTable::done() {
```

Comment:
interface done is required called in pipeline engine

Reply:
> interface done is required called in pipeline engine

I move it back since we don't have a better way